### DISCUSSION TITLE
Wait async subscriptions when I emit next value

### DISCUSSION THREAD

**Discussion by NaumovEvgeniy** [ReactiveX/rxjs#6701] [Q&A]
I have the next example
```
import { Subject } from "rxjs";

const subject = new Subject();

subject.subscribe(() => new Promise(res => {
	setTimeout(() => {
		console.log('!! 1');
		res(undefined)
	}, 500);
}))

subject.subscribe(() => new Promise(res => {
	setTimeout(() => {
		console.log('!! 2');
		res(undefined)
	}, 1000);
}))

console.log('>>> START')
subject.next(undefined);
console.log('<<< FINISH')

```

Console looks like

```
>>> START
<<< FINISH
!! 1
!! 2
I want the following behavious

>>> START
!! 1
!! 2
<<< FINISH
```

Can I reach expected behaviour or I should to use another aproach? I've digged into a source code, but I could not find solutions for me.

**Comment by voliva** [comment]
That's just.... imposible as you're doing it? What's your actual purpose?

Your code is equivalent to:

```js
console.log('>>> START')
new Promise(res => {
	setTimeout(() => {
		console.log('!! 1');
		res(undefined)
	}, 500);
})
new Promise(res => {
	setTimeout(() => {
		console.log('!! 2');
		res(undefined)
	}, 1000);
})
console.log('<<< FINISH')
```

Which will of course log 1 and 2 after "FINISH".

I can suggest something to "fix" it, which would be:

```js
import { Subject } from "rxjs";

const subject = new Subject();

subject.subscribe(() => console.log('!! 1'))
subject.subscribe(() => console.log('!! 2'))

console.log('>>> START')
subject.next(undefined);
console.log('<<< FINISH')
```

This will output what you want, but I'm sure it's not the answer you're looking for... What do you want to achieve?

**Reply by NaumovEvgeniy** [reply]
> What do you want to achieve?

In shortly, all subject's benefits like piping, subscribing, unsubscribing, etc.. Aslo I want to wait all async subscribers.

I develop my own event based system for my purposes. Maybe if you know some event listener framework based on rxjs, it will be great to suggest me
